core/base/iterator_factory.hpp

```diff
@@ -366,6 +366,130 @@ void swap(zip_iterator_reference<Iterators...> a,
 }
 
 
+template <typename IteratorType, typename PermuteFn>
+class permute_iterator {
+public:
+    using difference_type = std::ptrdiff_t;
+    using value_type = typename std::iterator_traits<IteratorType>::value_type;
+    using pointer = typename std::iterator_traits<IteratorType>::pointer;
+    using reference = typename std::iterator_traits<IteratorType>::reference;
+    using iterator_category = std::random_access_iterator_tag;
+
+    explicit permute_iterator() = default;
+
+    explicit permute_iterator(IteratorType it, PermuteFn perm)
+        : it_{std::move(it)}, idx_{}, perm_{std::move(perm)}
+    {}
+
+    permute_iterator& operator=(permute_iterator other)
+    {
+        it_ = other.it_;
+        idx_ = other.idx_;
+        // no perm_ = other.perm_ because lambdas are not copy-assignable
```

Comment:
I think just to be safe, I would prefer adding this line and using functor structs in the tests.

Reply:
Just for completion, lambdas have the copy assignment defined in c++20 (if they don't capture). Also, the copy constructor is still defined. So currently I'm hacking this by wrapping the lambda in a std::vector.

Reply:
We need copy assign, not copy construct. Lambdas have a default defined copy constructor, but their copy assign is deleted (unless in c++20), and we need here the assign. And by using `vector::emplace` I can use the copy constructor of the lambda.
I think using the vector results in better assembly (with function: https://godbolt.org/z/znG889nPj, with vector: https://godbolt.org/z/dxG7q7r4P) and since this might also be used by our omp backend, I would use the vector. I could also wrap the vector construction in a type (e.g. `copy_assignable`) to make it clearer what is happening.

Reply:
> Lambdas have a default defined copy constructor, but their copy assign is deleted (unless in c++20), and we need here the assign.

Thanks for the explanation, I didn't know that.
With the name `copy_assignable` and a comment mentioning that copy assign is deleted on lambdas, I am fine with your vector implementation.

However, note that your `std::function` implementation is very likely much faster than the `std::vector` implementation as you don't call `new` and `delete`.

Reply:
The new and delete is only once per iterator creation, but in the function case, you have function calls within the loop, which are not present in the vector case. But I will benchmark it to see if there are differences.

Reply:
To complete this, here is a quick benchmark: https://quick-bench.com/q/i8UiekPceUwmRSbAMOeBQEbzqvE (choose the bar visualization center right). The vector approach is indeed faster, but still takes ~2x than directly increasing the iterator by two.